pennylane_lightning/src/util/ConstantUtil.hpp

```diff
@@ -0,0 +1,179 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Contains utility functions for processing constants
+ */
+#pragma once
+#include <array>
+#include <cstdlib>
+#include <stdexcept>
+#include <tuple>
+
+namespace Pennylane::Util {
+/**
+ * @brief Lookup key in array of pairs. For a constexpr map-like behavior.
+ *
+ * @tparam Key Type of keys
+ * @tparam Value Type of values
+ * @tparam size Size of std::array
+ * @param arr Array to lookup
+ * @param key Key to find
+ */
+template <typename Key, typename Value, size_t size>
+constexpr auto lookup(const std::array<std::pair<Key, Value>, size> &arr,
+                      const Key &key) -> Value {
+    for (size_t idx = 0; idx < size; idx++) {
+        if (std::get<0>(arr[idx]) == key) {
+            return std::get<1>(arr[idx]);
+        }
+    }
+    throw std::range_error("The given key does not exist.");
+};
+
+/**
+ * @brief Check an array has an element.
+ *
+ * @tparam U Type of array elements.
+ * @tparam size Size of array.
+ * @param arr Array to check.
+ * @param elt Element to find.
+ */
+template <typename U, size_t size>
+constexpr auto array_has_elt(const std::array<U, size> &arr, const U &elt)
+    -> bool {
+    for (size_t idx = 0; idx < size; idx++) {
+        if (arr[idx] == elt) {
+            return true;
+        }
+    }
+    return false;
+};
+
+/**
+ * @brief Extract first elements from the array of pairs.
+ *
+ * @tparam T Type of the first elements.
+ * @tparam U Type of the second elements.
+ * @tparam size Size of the array.
+ * @param arr Array to extract.
+ */
+template <typename T, typename U, size_t size>
+constexpr std::array<T, size>
+first_elts_of(const std::array<std::pair<T, U>, size> &arr) {
+    // TODO: change to std::transform in C++20
+    std::array<T, size> res = {
+        T{},
+    };
+    for (size_t i = 0; i < size; i++) {
+        res[i] = std::get<0>(arr[i]);
+    }
+    return res;
+}
+/**
+ * @brief Extract second elements from the array of pairs.
+ *
+ * @tparam T Type of the first elements.
+ * @tparam U Type of the second elements.
+ * @tparam size Size of the array.
+ * @param arr Array to extract.
+ */
+template <typename T, typename U, size_t size>
+constexpr std::array<U, size>
+second_elts_of(const std::array<std::pair<T, U>, size> &arr) {
+    // TODO: change to std::transform in C++20
+    std::array<U, size> res = {
+        U{},
+    };
+    for (size_t i = 0; i < size; i++) {
+        res[i] = std::get<1>(arr[i]);
+    }
+    return res;
+}
```

Comment:
FYI, this function was the same as `fist_elts_of` previously. Besides these two functions, every other functions are the same.